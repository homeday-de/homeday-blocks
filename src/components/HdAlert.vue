<template>
  <div
    class="alert"
    :class="{
      [`alert--${type}`]: true,
      'alert--no-icon': noIcon,
    }"
  >
    <HdIcon
      v-if="!noIcon"
      :src="iconSrc"
      class="alert__icon"
    />
    <div
      class="alert__message"
    >
      <slot />
    </div>
  </div>
</template>

<script>
import HdIcon from 'homeday-blocks/src/components/HdIcon.vue';
import {
  checkCircleIcon as successIcon,
  infoIcon,
  warningIcon,
  errorIcon,
} from 'homeday-blocks/src/assets/small-icons';

// The first item is the default
export const TYPES = {
  info: {
    icon: infoIcon,
  },
  success: {
    icon: successIcon,
  },
  warning: {
    icon: warningIcon,
  },
  error: {
    icon: errorIcon,
  },
};

export default {
  name: 'HdAlert',
  components: {
    HdIcon,
  },
  props: {
    type: {
      type: String,
      default: Object.keys(TYPES)[0],
      validator(type) {
        return TYPES[type] !== undefined;
      },
    },
    icon: {
      type: String,
      default: '',
    },
    noIcon: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    iconSrc() {
      return this.icon || TYPES[this.type].icon;
    },
  },
};
</script>

<style lang="scss" scoped>
@import 'homeday-blocks/src/styles/mixins.scss';

.alert {
  $root: &;
  display: flex;
  padding: $inset-s;
  @include font('text-small');
  border-width: 1px;
  border-style: solid;
  border-radius: 3px;

  &--success {
    border-color: $success-color;
    color: $success-color;
    background-color: rgba($success-color, .07);
  }

  &--info {
    border-color: getShade($quaternary-color, 60);
    color: $quaternary-color;
  }

  &--warning {
    border-color: $tertiary-color;
    color: $tertiary-color;
    background-color: getShade($tertiary-color, 60);
  }

  &--error {
    border-color: $error-color;
    color: $error-color;
    background-color: rgba($error-color, .07);
  }

  &--no-icon {
    padding: $stack-s $inline-m;
  }

  &__icon {
    width: 24px;
    align-self: flex-start;
    margin-top: $stack-xxs;
    margin-right: $inline-s;

    ::v-deep path {
      #{$root}--success & {
        fill: $success-color;
      }
      #{$root}--info & {
        fill: $secondary-color;
      }
      #{$root}--warning & {
        fill: $tertiary-color;
      }
      #{$root}--error & {
        fill: $error-color;
      }
    }
  }
}
</style>
