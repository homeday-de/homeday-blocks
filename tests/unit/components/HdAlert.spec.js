import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import HdAlert, { TYPES } from '@/components/HdAlert.vue';

const defaultSlot = '<b>Default slot</b>';

const wrapperBuilder = wrapperFactoryBuilder(HdAlert, {
  slots: {
    default: defaultSlot,
  },
  shallow: true,
});

const MESSAGE_SELECTOR = '.alert__message';
const ICON_SELECTOR = '.alert__icon';

describe('HdAlert', () => {
  it('renders component', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders the slot correctly', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.find(MESSAGE_SELECTOR).element.innerHTML).toBe(defaultSlot);
  });

  Object.keys(TYPES).forEach((type) => {
    it(`should render alert--${type}`, () => {
      const wrapper = wrapperBuilder({
        props: {
          type,
        },
      });

      expect(wrapper.classes()).toContain(`alert--${type}`);
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  it('renders a custom icon', () => {
    const ICON_URL = 'https://placeholder.pics/svg/24x24';
    const wrapper = wrapperBuilder({
      props: {
        icon: ICON_URL,
      },
    });

    expect(wrapper.find(ICON_SELECTOR).attributes().src).toBe(ICON_URL);
  });

  it('doesn\'t render an icon if the prop `no-icon` is true', () => {
    const wrapper = wrapperBuilder({
      props: {
        noIcon: true,
      },
    });

    expect(wrapper.find(ICON_SELECTOR).exists()).toBe(false);
  });
});
