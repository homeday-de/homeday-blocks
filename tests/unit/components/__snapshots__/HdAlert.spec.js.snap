// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdAlert renders component 1`] = `
<div class="alert alert--info">
  <hdicon-stub src="" class="alert__icon"></hdicon-stub>
  <div class="alert__message"><b>Default slot</b></div>
</div>
`;
